### ISSUE TITLE
Gutenberg not retrieving latest featured image caption after modification

### ISSUE THREAD

**Issue by jfuenmayor96** [WordPress/gutenberg#24446]
**Describe the bug**
I noticed that whenever I select/upload a featured image in the post, and then I change its caption, the information I get after the modification is the old caption, not the modification I made. 

I have been able to reproduce this issue in both local and staging (a VPS) environments.

This has been happening to me since a couple of versions ago, since like gutenberg v5? I don't remember. It's been a long time.

Also, this is happening while using `withSelect` in a React component, not only in the web browser. 

**To reproduce**
1. Go to the post editor and create a new post.
2. Upload a new featured image, set a caption, and save the changes.
3. Write the following in the browser console to output the featured image caption: 
` wp.data.select( 'core' ).getMedia( wp.data.select('core/editor').getEditedPostAttribute('featured_media') ).caption; `
4. Open the featured image modal again, change the featured image, and write the command in step 3, and you will see the non-updated caption.

**Expected behavior**
After I select or upload a featured image, and then I change its caption, I expect Gutenberg to be able to retrieve the updated caption.

**Screenshots**
![Imgur](https://i.imgur.com/dLyQqx1.gif)

**Editor version:**
- WordPress version: 5.4.2
- Does the website has Gutenberg plugin installed, or is it using the block editor that comes by default? I have the Gutenberg plugin installed.
- If the Gutenberg plugin is installed, which version is it? 8.7.0

**Desktop:**
 - OS: Windows 10 Home Build 19041 - 64 Bits
 - Browser: Microsoft Edge 84.0.522.52. 
 - This happens with other browsers as well.

**Comment by ellatrix** [comment]
Sounds like the data is cached, perhaps in `getMedia`.

**Comment by jfuenmayor96** [comment]
@ellatrix I think so too. On my end, I made sure to rule out any cache problem. I've already deactivated the plugins that manage my cache.

**Comment by hawkmusician** [comment]
Having the same issue here. Info isn't updating till browser refresh (whether in browser or React component).

**Comment by Mamaduka** [comment]
The core data won't invalidate the cache after changes are made in the Media Library, since it has no way of knowing that. While we're still using the old Media Library, which isn't aware of the entities data store, keeping things in sync will be hard.

We could resolve the problem stated in the issue by invalidating the featured image cache if the user reselects the same image from the library.

@talldan, I think you were exploring a similar issue with media caches recently. What do you think, is it worth hotfixing this issue?

**Comment by talldan** [comment]
> @talldan, I think you were exploring a similar issue with media caches recently. What do you think, is it worth hotfixing this issue?

Yep, I tried a previous fix, but it was a bit hacky and was reverted due to another problem it caused.

I think technically it should be possible to listen to the backbone model's events, something like:
```jsx
wp.media.model.Attachments.all.on( 'change', () => // invalidate the attachments entity );
```

But I haven't found a good way to get that to work, plus it's not clear to me on which events we should invalidate which selector resolutions. Invalidating `invalidateResolution( 'getEntityRecords', [ 'postType', 'attachment' ] )` maybe? Then what about individual entity records (getEntityRecord), and which exact args to invalidate. I think the API is a bit hard to work with from the external perspective.

Anyway, I think this might be quite low priority given there aren't many places in core where it'll actually cause a bug. Plugins can always add their own code to fix this if they feel inclined or contribute a fix.

**Comment by Mamaduka** [comment]
Thanks, @talldan!

Agree, it's hard to find the right balance of what to invalidate. Doing it for the whole entity `kind/name` could trigger a lot of refetching based on how media-heavy the website/content is.

I think consumers/plugins can handle this better based on their needs and requirements. Considering that and the low activity on this issue, I would suggest closing this as "maybe later".
